### PR TITLE
refactor: deduplicate the message sign/verify code

### PR DESCRIPTION
**Pull request by vasild**
The message signing and verifying logic was replicated in a few places
in the code. Consolidate in a newly introduced `MessageSign()` and
`MessageVerify()` and add unit tests for them.

### CODE REVIEW

File: src/util/message.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#include <serialize.h>       // For SER_GETHASH
+#include <util/message.h>
+#include <util/strencodings.h> // For DecodeBase64()
+
+// We moved a call to boost::get() here from another file. It is a good practice
+// to include the header that provides it (<boost/variant/get.hpp>). However
+// test/lint/lint-includes.sh:90 complains that a new Boost dependency is
+// introduced because even though boost::get() is used in many other places in
+// the code, nobody has bothered to include <boost/variant/get.hpp>.
+//#include <boost/variant/get.hpp> // For boost::get()
```
Comment (vasild) [inline src/util/message.cpp]: For some reason it is possible to use `boost::get()` without including the header that provides it. Probably included from somewhere else, maybe from QT stuff.

Just drop lines 15-20 or somehow silence the lint script to allow this header to be included (a new boost dependency is not added by this patch)?

**Comment by fanquake** [comment]
Looks like this is going to overlap with #17557.

**Comment by vasild** [comment]
Affairs with conflicting PRs:

* #17557 util: Refactor message hashing into a utility function
Both PRs target the same duplicated code! However this PR (#17577) has broader scope as it deduplicates more code.

* #17493 util: Forbid ambiguous repeated assignments in config file
Purely mechanical conflicts. Both PRs touch neighboring code. Either one can be merged first, the second will need to be adjusted (easy).

* #17473 refactor: Settings code cleanups
Purely mechanical conflicts. Same as above.

* #17399 validation: Templatize ValidationState instead of subclassing
Purely mechanical conflicts. Same as above.

* #16545 Implement missing error checking for ArgsManager flags
Purely mechanical conflicts. Same as above.

**Comment by vasild** [comment]
@jkczyz What do you think about this PR and its relation to #17557? Merge both in one PR?

**Comment by jkczyz** [comment]
> @jkczyz What do you think about this PR and its relation to #17557? Merge both in one PR?

I have #17557 based on an outstanding PR (#17399), though I suppose it didn't have to be. The former resulted from a comment in the latter. I deliberately kept the change small so it could hopefully be merged quicker. Though I wanted to add a test to clarify why the "magic" string is needed.

That said, from a design perspective, I prefer your PR as it hides the complexity around signing a message.

**Comment by vasild** [comment]
@jkczyz thanks for the feedback! I did cherry-pick the last two commits from https://github.com/bitcoin/bitcoin/pull/17557 into this PR under e3d85bcd5 - the hashing utility function and the unit test for it (to clarify why the "magic" string is needed).

**Comment by vasild** [comment]
> ### Conflicts
> 
> Reviewers, this pull request conflicts with the following ones:
> 
>     * [#17593](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17593.html) (test: move more utility functions into test utility library by mzumsande)
> 
>     * [#17581](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17581.html) (refactor: Remove settings merge reverse precedence code by ryanofsky)
> 
>     * [#17580](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17580.html) (refactor: Add ALLOW_LIST flags and enforce usage in CheckArgFlags by ryanofsky)
> 
>     * [#17493](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17493.html) (util: Forbid ambiguous multiple assignments in config file by ryanofsky)
> 
>     * [#17473](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17473.html) (refactor: Settings code cleanups by ryanofsky)
> 
>     * [#17399](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17399.html) (validation: Templatize ValidationState instead of subclassing by jkczyz)
> 
>     * [#16545](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/16545.html) (refactor: Implement missing error checking for ArgsManager flags by ryanofsky)

All of the above conflicts are mechanical, small and easy to resolve.

**Comment by instagibbs** [comment]
concept ACK, though I'd probably be more pedantic in calling it "Bitcoin Signed Message Protocol" or something. Unfortunately there's no BIP to refer to.

File: src/util/message.cpp (hunk 1 above)
Comment (vasild) [reply src/util/message.cpp]: I dropped lines 15-20

**Comment by vasild** [comment]
@instagibbs Thanks for the review! Where would you put the "Bitcoin Signed Message Protocol" label?

We do message signing and verification now. What about encryption and decryption? Would that be useful feature or just a bloat that is rarely used? If encryption may be added in the future then maybe just "Bitcoin Message Protocol" would be a better name, or "Bitcoin Text Message Protocol".

**Comment by laanwj** [comment]
I don't think encryption / decryption in bitcoind is a good idea. It is best to avoid using keys for multiple purposes (e.g. use a signing key for encryption, and vice versa). Also it's scope creep: Bitcoin core is a validating full node. It should do that well, and not try to make a GPG-ish mess of it that tries to do everything with a mash-up of features but nothing particularly well…

**Comment by Sjors** [comment]
See also @achow101's #18115

**Comment by vasild** [comment]
> See also @achow101's #18115

@Sjors thanks for the notice!

This PR deduplicates message signing and message verification code.

I peeked at https://github.com/bitcoin/bitcoin/pull/18115/ and among other things it deduplicates the message signing code from 2 of 3 places and does not deduplicate the message verification code. Assuming it also fixes the [3rd place](https://github.com/bitcoin/bitcoin/pull/18115/#issuecomment-585122576) where message signing code is repeated, then it will achieve half of this PR.

Should https://github.com/bitcoin/bitcoin/pull/18115/ hit `master` before this PR, then I will trim this PR to only consolidate the message verification code.

**Comment by vasild** [comment]
> I peeked at #18115 and among other things it deduplicates the message signing code from 2 of 3 places and does not deduplicate the message verification code. Assuming it also fixes the 3rd place...

That was fixed and now #18115 deduplicates all 3 message signing repetitions.

[MessageSign()](https://github.com/bitcoin/bitcoin/pull/17577/files#diff-f7b16fbe87fc15c6f15347e78210e501R24) from https://github.com/bitcoin/bitcoin/pull/17577 is equivalent to [SignMessage()](https://github.com/bitcoin/bitcoin/pull/18115/files#diff-d546634098d8003b6397c8643b74e47dR520) from https://github.com/bitcoin/bitcoin/pull/18115. The former has unit tests added in https://github.com/bitcoin/bitcoin/pull/17577.

File: src/util/message.cpp [hunk 2]
```diff
@@ -14,7 +15,27 @@
 #include <string>
 #include <vector>
 
-const std::string strMessageMagic = "Bitcoin Signed Message:\n";
+static const std::string MESSAGE_MAGIC = "Bitcoin Signed Message:\n";
+
+bool MessageSign(
+    const CKey& privkey,
+    const std::string& message,
+    std::string* signature)
```
Comment (achow101) [inline src/util/message.cpp]: This should be a reference (`std::string&`) instead of pointer.

File: src/test/util_tests.cpp [hunk 3]
```diff
@@ -2094,4 +2096,13 @@ BOOST_AUTO_TEST_CASE(message_verify)
         MessageVerificationResult::OK);
 }
 
+BOOST_AUTO_TEST_CASE(message_hash)
+{
+    const std::string unsigned_tx = "...";
+    const uint256 signature_hash = Hash(unsigned_tx.begin(), unsigned_tx.end());
+
+    const uint256 message_hash = MessageHash(unsigned_tx);
+    BOOST_CHECK_NE(message_hash, signature_hash);
```
Comment (achow101) [inline src/test/util_tests.cpp]: I think you should have on additional check here with the full hashed message. i.e. also hash `Bitcoin Signed Message:\n...` and compare that with `message_hash`.

File: src/util/message.cpp [hunk 4]
```diff
@@ -14,7 +15,27 @@
 #include <string>
 #include <vector>
 
-const std::string strMessageMagic = "Bitcoin Signed Message:\n";
+static const std::string MESSAGE_MAGIC = "Bitcoin Signed Message:\n";
```
Comment (achow101) [inline src/util/message.cpp]: Renaming this should probably go in the next commit?

**Review by achow101 — COMMENTED** [review]
Couple of nits. otherwise ACK 1a00fb84171666451cb5a21ba454755dab6e0328

File: src/qt/signverifymessagedialog.cpp [hunk 5]
```diff
@@ -189,51 +189,55 @@ void SignVerifyMessageDialog::on_addressBookButton_VM_clicked()
...
-        ui->statusLabel_VM->setStyleSheet("QLabel { color: red; }");
-        ui->statusLabel_VM->setText(tr("The signature did not match the message digest.") + QString(" ") + tr("Please check the signature and try again."));
-        return;
+    const std::string& address = ui->addressIn_VM->text().toStdString();
+    const std::string& signature = ui->signatureIn_VM->text().toStdString();
+    const std::string& message = ui->messageIn_VM->document()->toPlainText().toStdString();
+
+    switch (MessageVerify(address, signature, message)) {
+        case MessageVerificationResult::ERR_INVALID_ADDRESS:
+            ui->statusLabel_VM->setStyleSheet("QLabel { color: red; }");
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/qt/signverifymessagedialog.cpp]: Done.

File: src/util/message.cpp (hunk 2 above)
Comment (vasild) [reply src/util/message.cpp]: Done.

File: src/test/util_tests.cpp (hunk 3 above)
Comment (vasild) [reply src/test/util_tests.cpp]: Done.

File: src/util/message.cpp (hunk 4 above)
Comment (vasild) [reply src/util/message.cpp]: Done.

File: src/test/util_tests.cpp (hunk 3 above)
Comment (vasild) [reply src/test/util_tests.cpp]: Notice: assuming `CHashWriter hasher(SER_GETHASH, 0);`, then:
```
hasher << "ab" << "cd";
```
would produce a different result than:
```
hasher << "abcd";
```

**Comment by Sjors** [comment]
tACK ad39c8fd8f2f7752e4f4508ab880873008c7452e, thanks for adding tests as well

**Comment by achow101** [comment]
I'm seeing a new compiler warning:
```
In file included from ./script/signingprovider.h:13,
                 from ./outputtype.h:10,
                 from rpc/misc.cpp:8:
rpc/misc.cpp: In function ‘UniValue verifymessage(const JSONRPCRequest&)’:
./sync.h:179:104: warning: control reaches end of non-void function [-Wreturn-type]
  179 | #define LOCK(cs) DebugLock<decltype(cs)> PASTE2(criticalblock, __COUNTER__)(cs, #cs, __FILE__, __LINE__)
      |                                                                                                        ^
rpc/misc.cpp:273:5: note: in expansion of macro ‘LOCK’
  273 |     LOCK(cs_main);
      |     ^~~~
```

File: src/rpc/misc.cpp [hunk 6]
```diff
@@ -276,31 +276,19 @@ static UniValue verifymessage(const JSONRPCRequest& request)
...
+        case MessageVerificationResult::ERR_ADDRESS_NO_KEY:
+            throw JSONRPCError(RPC_TYPE_ERROR, "Address does not refer to key");
+        case MessageVerificationResult::ERR_MALFORMED_SIGNATURE:
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Malformed base64 encoding");
+        case MessageVerificationResult::ERR_PUBKEY_NOT_RECOVERED:
+        case MessageVerificationResult::ERR_NOT_SIGNED:
+            return false;
+        case MessageVerificationResult::OK:
+            return true;
     }
```
Comment (achow101) [inline src/rpc/misc.cpp]: You need another `return false` statement after this `switch`.

Comment (Sjors) [reply src/rpc/misc.cpp]: Nice catch. I'm not getting that `-Wreturn-type` warning on macOS. Only one of the Travis machines noticed it: https://travis-ci.org/bitcoin/bitcoin/jobs/649943882?utm_medium=notification&utm_source=github_status

Comment (vasild) [reply src/rpc/misc.cpp]: Added `return false;`, thanks!

**Comment by Sjors** [comment]
re-ACK e193a84fb28068e38d5f54fbfd6208428c5bb655

**Comment by achow101** [comment]
ACK e193a84fb28068e38d5f54fbfd6208428c5bb655

**Comment by Sjors** [comment]
@meshcollider some of this touches the wallet - and #18115 that builds on top of this definitely does - so maybe you can take a look?

File: src/util/message.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    //! The message verification was successful.
+    OK
+};
+
+/** Verify a signed message.
+ * @param[in] address Signer's bitcoin address, it must refer to a public key.
+ * @param[in] signature The signature in base64 format.
+ * @param[in] message The message that was signed.
+ * @return result code */
+MessageVerificationResult MessageVerify(
```
Comment (instagibbs) [inline src/util/message.h]: nit: I'd prefer actual types used instead of strings for all args.

**Review by instagibbs — COMMENTED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/17577/commits/e193a84fb28068e38d5f54fbfd6208428c5bb655

File: src/util/message.h (hunk 7 above)
Comment (instagibbs) [reply src/util/message.h]: this is a non-blocking comment :) let's get this merged

**Review by meshcollider — COMMENTED** [review]
utACK e193a84fb28068e38d5f54fbfd6208428c5bb655

Nice cleanup, thanks!

**Pull request merged**